src/miner.erl

```diff
@@ -600,9 +445,229 @@ terminate(Reason, _State) ->
+                         {ok, 2} -> blockchain_txn_rewards_v2;
+                         _       -> blockchain_txn_rewards_v1
+                     end,
+                {ok, Rewards} = RewardsMod:calculate_rewards(Start, End, Chain),
+                lager:debug("RewardsMod: ~p, Rewards: ~p~n", [RewardsMod, Rewards]),
+                RewardsTxn = RewardsMod:new(Start, End, Rewards),
+                %% To cut down on the size of group txn blocks, which we'll
+                %% need to fetch and store all of to validate snapshots, we
+                %% discard all other txns for this block.
+                Transactions =
+                    lists:sort(
+                        %% TODO Rename blockchain_txn:sort to blockchain_txn:(compare|cmp)
+                        fun blockchain_txn:sort/2,
+                        [RewardsTxn, ConsensusGroupTxn]
+                     ),
+                {Epoch, BlockHeightNext, Transactions, InvalidTransactions};
+            _ ->
+                {ElectionEpoch0, EpochStart0, ValidTransactions, InvalidTransactions}
+        end.
+
+-spec txn_is_rewards(blockchain_txn:txn()) -> boolean().
+txn_is_rewards(Txn) ->
+    Rewards = [blockchain_txn_rewards_v1, blockchain_txn_rewards_v2],
+    lists:member(blockchain_txn:type(Txn), Rewards).
+
+-spec metadata_to_seen_bbas(metadata()) ->
+    [{{pos_integer(), binary()}, binary()}].
+metadata_to_seen_bbas(M) ->
+    [{{J, S}, B} || {J, #{seen := S, bba_completion := B}} <- metadata_only_v2(M)].
+
+-spec metadata_only_v2(metadata()) ->
+    [{non_neg_integer(), metadata_v2()}].
+metadata_only_v2(Metadata) ->
+    lists:filter(fun ({_, M}) -> is_map(M) end, Metadata).
+
+-spec meta_to_stamp_hashes(metadata()) ->
+    {
+        Stamps :: [integer()],
+        Hashes :: [blockchain_block:hash()]
+    }.
+meta_to_stamp_hashes(Metadata) ->
+    lists:unzip([metadata_as_v1(M) || {_, M} <- Metadata]).
+
+-spec metadata_as_v1(metadata_v1() | metadata_v2()) -> metadata_v1().
+metadata_as_v1(#{head_hash := H, timestamp := S}) -> {S, H}; % v2 -> v1
+metadata_as_v1({S, H})                            -> {S, H}. % v1 -> v1
+
+-spec snapshot_hash(L, H, M, V) -> binary()
+    when L :: blockchain_ledger_v1:ledger(),
+         H :: non_neg_integer(),
+         M :: metadata(),
+         V :: non_neg_integer().
+snapshot_hash(Ledger, BlockHeightNext, Metadata, VotesNeeded) ->
+    %% Find a snapshot hash.  If not enabled or we're unable to determine or
+    %% agree on one, just leave it blank, so other nodes can absorb it.
+    case blockchain:config(?snapshot_interval, Ledger) of
+        {ok, Interval} when (BlockHeightNext - 1) rem Interval == 0 ->
+            Hashes = [H || {_, #{snapshot_hash := H}} <- metadata_only_v2(Metadata)],
+            pick_consented_or_default(VotesNeeded, Hashes, <<>>);
+        _ ->
+            <<>>
+    end.
+
+-spec pick_consented_or_default(non_neg_integer(), [X], X) -> X.
```

Comment:
name is odd? `most_common_over_threshold` maybe?

Reply:
`most_common_or_default`?

Reply:
`common_enough_or_default`?

Reason why I'm stuck on picking `default` over `threshold` in the name: threshold seems reasoably-obvious in relation to commonality, whereas falling back on default is not.